dbms/src/Flash/Mpp/MPPTunnel.cpp

```diff
@@ -176,196 +167,229 @@ void MPPTunnelBase<Writer>::write(const mpp::MPPDataPacket & data, bool close_af
     }
     else
     {
         LOG_FMT_TRACE(log, "start waitUntilConnectedOrFinished");
-        cv_for_connected_or_finished.wait(lk, connected_or_finished);
+        cv_for_status_changed.wait(lk, not_unconnected);
         LOG_FMT_TRACE(log, "end waitUntilConnectedOrFinished");
     }
-    if (!connected)
+    if (status == TunnelStatus::Unconnected)
         throw Exception("MPPTunnel can not be connected because MPPTask is cancelled");
 }
 
-template <typename Writer>
-void MPPTunnelBase<Writer>::consumerFinish(const String & err_msg, bool need_lock)
+StringRef MPPTunnel::statusToString()
+{
+    switch (status)
+    {
+    case TunnelStatus::Unconnected:
+        return "Unconnected";
+    case TunnelStatus::Connected:
+        return "Connected";
+    case TunnelStatus::WaitingForSenderFinish:
+        return "WaitingForSenderFinish";
+    case TunnelStatus::Finished:
+        return "Finished";
+    default:
+        RUNTIME_ASSERT(false, log, "Unknown TaskStatus {}", status);
+    }
+}
+
+void TunnelSender::consumerFinish(const String & err_msg)
 {
-    // must finish send_queue outside of the critical area to avoid deadlock with write.
     LOG_FMT_TRACE(log, "calling consumer Finish");
-    send_queue.finish();
-    auto rest_work = [this, &err_msg] {
-        // it's safe to call it multiple times
-        if (finished && consumer_state.errHasSet())
+    send_queue->finish();
+    // it's safe to call it multiple times
+    {
+        std::unique_lock lk(state_mu);
+        if (consumer_state.msgHasSet())
             return;
-        finished = true;
-        // must call setError in the critical area to keep consistent with `finished` from outside.
-        consumer_state.setError(err_msg);
-        cv_for_connected_or_finished.notify_all();
-    };
-    if (need_lock)
+        consumer_state.setMsg(err_msg);
+    }
+}
+
+SyncTunnelSender::~SyncTunnelSender()
+{
+    LOG_FMT_TRACE(log, "waiting child thread finished!");
+    thread_manager->wait();
+}
+
+void SyncTunnelSender::sendJob()
+{
+    GET_METRIC(tiflash_thread_count, type_active_threads_of_establish_mpp).Increment();
+    GET_METRIC(tiflash_thread_count, type_max_threads_of_establish_mpp).Set(std::max(GET_METRIC(tiflash_thread_count, type_max_threads_of_establish_mpp).Value(), GET_METRIC(tiflash_thread_count, type_active_threads_of_establish_mpp).Value()));
+    String err_msg;
+    try
     {
-        std::unique_lock lk(mu);
-        rest_work();
+        MPPDataPacketPtr res;
+        while (send_queue->pop(res))
+        {
+            if (!writer->write(*res))
+            {
+                err_msg = "grpc writes failed.";
+                break;
+            }
+        }
     }
-    else
-        rest_work();
+    catch (...)
+    {
+        err_msg = getCurrentExceptionMessage(true);
+    }
+    if (!err_msg.empty())
+    {
+        err_msg = fmt::format("{} meet error: {}", tunnel_id, err_msg);
+        LOG_ERROR(log, err_msg);
```

Comment:
I suggest to trim the stack trace information in `err_msg` after it is logged. Otherwise the error message will be too long and hard to read. Please refer to #5304 for the details

Reply:
Got it!

Reply:
Got it, nice suggestions!